File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.js

var Trie = function() {
    this.words = {};
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let current = this.words;
    for(let char of word){
        if(!current[char]) current[char] = {};
        current = current[char]
    }
    current['word?'] = true;
    
};

/** 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    let current = this.words;
    for(let char of word){
        if(!current[char]) return false;
        current = current[char]
    }
    return !!current['word?']
};

/** 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    let current = this.words;
    for(let char of prefix){
        if(!current[char]) return false;
        current = current[char]
    }
    return !!current
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.py
class Trie:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        current = self.root
        
        for char in word:
            i = ord(char)-ord('a')
            if not current.children[i]: current.children[i] = Node()
            current = current.children[i]
        
        current.word = True

    def search(self, word: str) -> bool:
        current = self.root
        
        for char in word:
            i = ord(char)-ord('a')
            if not current.children[i]: return False
            current = current.children[i]

        return current.word

    def startsWith(self, prefix: str) -> bool:
        current = self.root
        
        for char in prefix:
            i = ord(char)-ord('a')
            if not current.children[i]: return False
            current = current.children[i]

        return current != None
        

class Node:
    def __init__(self):
        self.word = False
        self.children = [None]*26